backend/src/main/java/com/rota/auth/AuthenticationUtils.java

```diff
@@ -1,51 +1,219 @@
 package com.rota.auth;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
 import com.rota.database.orm.staff.Role;
+import com.rota.exceptions.AuthenticationHttpException;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.Charset;
 import java.util.Optional;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.json.JsonParseException;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Service;
 
 /**
- * Dummy class which represents user authentication.
+ * Utils for handling authorisation tokens and general authentication related methods.
  */
+@Service
 public class AuthenticationUtils {
+
+  @Value("${auth0.managementapi.client_secret}")
+  private String clientSecret;
+
+  @Value("${auth0.managementapi.client_id}")
+  private String clientId;
+
+  @Value("${auth0.managementapi.audience}")
+  private String audience;
+
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+
+  private final HttpClient httpClient = HttpClient.newBuilder()
+      .version(HttpClient.Version.HTTP_2)
+      .build();
+
   /**
-   * Checks if the provided token is valid.
-   * Only returns false if token is "false" (case insensitive).
-   * @param token Authentication token.
-   * @return Whether authentication was successful.
+   * TODO decide if it is definitely necessary to check with Auth0. Another method for that perhaps?
+   * Checks if the current token is valid locally and via Auth0.
+   *
+   * @return Whether the current token is valid.
    */
-  public static boolean validateToken(String token) {
-    // This will only return false if the token is "false" (case insensitive)
-    return !token.toLowerCase().equals("false");
+  public boolean validateToken(String token) {
+    HttpRequest request = HttpRequest.newBuilder()
+        .GET()
+        .uri(URI.create("https://rota-flex-101.eu.auth0.com/userinfo"))
+        .setHeader("Authorization", "Bearer " + token)
+        .build();
+
+    HttpResponse<String> response;
+    try {
+      response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+    } catch (IOException | InterruptedException e) {
```

Comment:
This piece of code is repeated a few times so it should be refactored as a separate function.
We may want to convert AuthenticationUtils to Authentication class and keep the response as a class variable. That way we won't have to send requests to auth0 over and over again.

Reply:
Yeah good point, but by sending them again we validate that the token hasn't expired and is valid. Should definitely be looked into though.

Reply:
Can we extract the expiration time and check it ourselves?
I think limiting the traffic with auth0 would be good for our performance (and maybe costs once we're on google cloud).

Reply:
I've just been testing expiration stuff. The built in `SecurityContextHandler` seems to manage this itself and the token is invalid if expired. I tested this by setting the expiration time as 60 seconds and making repeated calls to the API with the same token.

The token stopped working after the expiration time (although it wasn't exact, was sometimes a few seconds late?).

This means that we can rely on `SecurityContextHolder` to only accept valid, non expired tokens for the current execution.